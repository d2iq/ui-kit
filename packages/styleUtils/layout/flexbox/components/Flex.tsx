import * as React from "react";
import { css } from "@emotion/css";
import { SpaceSize } from "../../../../shared/styles/styleUtils/modifiers/modifierUtils";
import {
  FlexboxProperties,
  flex,
  applyFlexItemGutters
} from "../../../../shared/styles/styleUtils/layout/flexbox";

interface FlexProps extends FlexboxProperties {
  className?: string;
  /**
   * The size of the space between each `FlexItem` child. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  gutterSize?: SpaceSize;
  /**
   * Which HTML tag to render the component with
   */
  tag?: keyof React.ReactHTML;
}

const Flex: React.FC<FlexProps> = ({
  children,
  className,
  gutterSize,
  tag = "div",
  ...flexboxProperties
}) => {
  const FlexEl = tag;

  return (
    <FlexEl
      className={css`
        ${flex({ ...flexboxProperties })};
        ${applyFlexItemGutters(flexboxProperties.direction, gutterSize)};
        ${className};
      `}
      data-cy="flex"
    >
      {children}
    </FlexEl>
  );
};

Flex.defaultProps = {
  align: "flex-start",
  direction: "row",
  wrap: "nowrap",
  justify: "flex-start",
  gutterSize: "none"
};

export default Flex;
