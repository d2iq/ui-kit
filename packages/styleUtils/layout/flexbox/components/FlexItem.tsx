import * as React from "react";
import { css } from "@emotion/css";
import {
  atMediaUp,
  BreakpointConfig
} from "../../../../shared/styles/breakpoints";
import { getResponsiveStyle } from "../../../../shared/styles/styleUtils";
import { flexItem } from "../../../../shared/styles/styleUtils/layout/flexbox";

type FlexStrategy = "shrink" | "grow";

interface FlexItemProps {
  className?: string;
  /**
   * Whether the flex item should grow or shrinkwrap to it's children. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  flex?: BreakpointConfig<FlexStrategy>;
  /**
   * The rate at which a flex item should grow. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  growFactor?: BreakpointConfig<React.CSSProperties["flexGrow"]>;
  /**
   * The position at which a flex item should be visually rendered. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  order?: BreakpointConfig<React.CSSProperties["order"]>;
  /**
   * Which HTML tag to render the component with
   */
  tag?: keyof React.ReactHTML;
}

const getResponsiveFlexItemStyles = flexVal =>
  typeof flexVal !== "object"
    ? flexItem(flexVal)
    : Object.keys(flexVal).map(breakpoint =>
        atMediaUp[breakpoint](flexItem(flexVal[breakpoint]))
      );

const FlexItem: React.FC<FlexItemProps> = ({
  children,
  className,
  flex,
  growFactor,
  order,
  tag = "div"
}) => {
  const FlexItemEl = tag;

  return (
    <FlexItemEl
      className={css`
        ${getResponsiveFlexItemStyles(flex)};
        ${getResponsiveStyle("flex-grow", growFactor)};
        ${getResponsiveStyle("order", order)};
        ${className};
      `}
      data-cy="flexItem"
    >
      {children}
    </FlexItemEl>
  );
};

FlexItem.defaultProps = {
  flex: "grow"
};

export default FlexItem;
