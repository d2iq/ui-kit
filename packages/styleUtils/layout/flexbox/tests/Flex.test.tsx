import React from "react";
import { shallow } from "enzyme";
import { createSerializer } from "@emotion/jest";
import toJson from "enzyme-to-json";

import Flex from "../components/Flex";
import FlexItem from "../components/FlexItem";

expect.addSnapshotSerializer(createSerializer());

describe("Flex", () => {
  it("renders default", () => {
    const component = shallow(
      <Flex>
        <FlexItem>content</FlexItem>
        <FlexItem>content</FlexItem>
      </Flex>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with all props", () => {
    const component = shallow(
      <Flex
        align="center"
        justify="center"
        direction="column"
        wrap="nowrap"
        gutterSize="l"
      >
        <FlexItem>content</FlexItem>
        <FlexItem>content</FlexItem>
      </Flex>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with responsive props", () => {
    const component = shallow(
      <Flex
        align={{ default: "flex-start", medium: "center" }}
        justify={{ default: "flex-start", medium: "center" }}
        direction={{ default: "column", medium: "row" }}
        wrap={{ default: "wrap", medium: "nowrap" }}
        gutterSize={{ default: "s", medium: "m" }}
      >
        <FlexItem>content</FlexItem>
        <FlexItem>content</FlexItem>
      </Flex>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
});
