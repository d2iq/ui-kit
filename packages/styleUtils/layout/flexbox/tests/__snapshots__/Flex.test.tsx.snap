// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex renders default 1`] = `
.emotion-0 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  height: auto;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  -webkit-column-gap: 0;
  column-gap: 0;
  row-gap: 0;
}

.emotion-0>div {
  width: auto;
}

@supports (column-gap: 1px) and (row-gap: 1px) {
  .emotion-0>*:not(:last-child) {
    padding-right: 0;
    padding-bottom: 0;
  }
}

<div
  className="emotion-0"
  data-cy="flex"
>
  <FlexItem
    flex="grow"
  >
    content
  </FlexItem>
  <FlexItem
    flex="grow"
  >
    content
  </FlexItem>
</div>
`;

exports[`Flex renders with all props 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
  -webkit-column-gap: 0;
  column-gap: 0;
  row-gap: 24px;
}

.emotion-0>div {
  width: 100%;
}

@supports (column-gap: 1px) and (row-gap: 1px) {
  .emotion-0>*:not(:last-child) {
    padding-right: 0;
    padding-bottom: 24px;
  }
}

<div
  className="emotion-0"
  data-cy="flex"
>
  <FlexItem
    flex="grow"
  >
    content
  </FlexItem>
  <FlexItem
    flex="grow"
  >
    content
  </FlexItem>
</div>
`;

exports[`Flex renders with responsive props 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    height: 100%;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    height: auto;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-box-flex-wrap: wrap;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-box-flex-wrap: nowrap;
    -webkit-flex-wrap: nowrap;
    -ms-flex-wrap: nowrap;
    flex-wrap: nowrap;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-box-pack: start;
    -ms-flex-pack: start;
    -webkit-justify-content: flex-start;
    justify-content: flex-start;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
  }
}

@media (min-width: 0) {
  .emotion-0>div {
    width: 100%;
  }
}

@media (min-width: 900px) {
  .emotion-0>div {
    width: auto;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-column-gap: 0;
    column-gap: 0;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-column-gap: 16px;
    column-gap: 16px;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    row-gap: 12px;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    row-gap: 0;
  }
}

@supports (column-gap: 1px) and (row-gap: 1px) {
  @media (min-width: 0) {
    .emotion-0>*:not(:last-child) {
      padding-right: 0;
    }
  }

  @media (min-width: 900px) {
    .emotion-0>*:not(:last-child) {
      padding-right: 16px;
    }
  }

  @media (min-width: 0) {
    .emotion-0>*:not(:last-child) {
      padding-bottom: 12px;
    }
  }

  @media (min-width: 900px) {
    .emotion-0>*:not(:last-child) {
      padding-bottom: 0;
    }
  }
}

<div
  className="emotion-0"
  data-cy="flex"
>
  <FlexItem
    flex="grow"
  >
    content
  </FlexItem>
  <FlexItem
    flex="grow"
  >
    content
  </FlexItem>
</div>
`;
